### PR TITLE
Msalem/enhance deployment process

### PR DESCRIPTION
**Pull request by msalemcode**
1- Adding New deployment script to speed up the process by dropping ARM template dependency
1- Adding deployment script to deploy KV with SaaS Accelerator and secure the configuration secrets

### CODE REVIEW

File: deployment/Templates/Deploy.ps1 [hunk 1]
```diff
@@ -186,117 +186,86 @@ if($LogoURLico) {
...
-   bacpacUrl                    = "$URLToBacpacFromStorage"
-   SAASKeyForbacpac             = ""
-   PublisherAdminUsers          = "$PublisherAdminUsers"
-   PathToWebApplicationPackages = "$PathToWebApplicationPackages"
-   MeteredSchedulerSupport      = "$MeteredSchedulerSupport"
-}
-
-
 # Create RG if not exists
 New-AzResourceGroup -Name $ResourceGroupForDeployment -Location $location -Force
```
Comment (alexdrenea) [inline deployment/Templates/Deploy.ps1:219]: we could be using az cli to deploy this as well. Since we're using it below to deploy the rest of resources.

File: deployment/Templates/Deploy.ps1 [hunk 2]
```diff
@@ -186,117 +186,86 @@ if($LogoURLico) {
...
+$WebAppNameService=$WebAppNamePrefix+"AmpSvcPlan"
+$WebAppNameAdmin=$WebAppNamePrefix+"-admin"
+$WebAppNamePortal=$WebAppNamePrefix+"-portal"
+$Connection="Data Source=tcp:"+$SQLServerName+".database.windows.net,1433;Initial Catalog=AMPSaaSDB;User Id="+$SQLAdminLogin+"@"+$SQLServerName+".database.windows.net;Password="+$SQLAdminLoginPassword+";"
+az appservice plan create -g $ResourceGroupForDeployment -n $WebAppNameService --sku B1
+
+
+Write-host "📜  Create publisher Admin webapp"
+az webapp create -g $ResourceGroupForDeployment -p $WebAppNameService -n $WebAppNameAdmin  --runtime dotnet:6
+az webapp config connection-string set -g $ResourceGroupForDeployment -n $WebAppNameAdmin -t mysql --settings DefaultConnection=$Connection
```
Comment (alexdrenea) [inline deployment/Templates/Deploy.ps1:250]: Should the type be "SQLAzure" instead of mysql?

**Comment by alexdrenea** [comment]
Love seeing this making it into the accelerator. Hope to see this approved soon!

File: deployment/Templates/Deploy.ps1 [hunk 3]
```diff
@@ -186,117 +186,86 @@ if($LogoURLico) {
...
-New-AzStorageContainer -Name $ContainerName -Context $ctx -Permission Blob 
-Set-AzStorageBlobContent -File $LocalPathToBacpacFile -Container $ContainerName -Blob $BlobName -Context $ctx -Force
-
-$URLToBacpacFromStorage = (Get-AzStorageBlob -blob $BlobName -Container $ContainerName -Context $ctx).ICloudBlob.uri.AbsoluteUri
-
-Write-host "📜  Uploaded the bacpac file to $URLToBacpacFromStorage"    
-
+Write-host "☁  Prepare publish files for the web application"
 
 if (!($MeteredSchedulerSupport))
```
Comment (alexdrenea) [inline deployment/Templates/Deploy.ps1:191]: This parameter could be a Boolean instead of string
`[bool][Parameter()]$MeteredSchedulerSupport`

and then it can be queried with 
`if ($MeteredSchedulerSupport -eq $true)`

When calling the script, set this to true by just specifying it in the command (-MeteredSchedulerSupport) or have it as false by omitting it.

File: deployment/Templates/Deploy.ps1 [hunk 4]
```diff
@@ -186,117 +186,86 @@ if($LogoURLico) {
...
-
 # Create RG if not exists
 New-AzResourceGroup -Name $ResourceGroupForDeployment -Location $location -Force
 
 Write-host "📜  Deploying the ARM template to set up resources"
-# Deploy resources using ARM template
-New-AzResourceGroupDeployment -ResourceGroupName $ResourceGroupForDeployment -TemplateFile $PathToARMTemplate -TemplateParameterObject $ARMTemplateParams
 
+Write-host "Create SQL Server"
+az sql server create --name $SQLServerName --resource-group $ResourceGroupForDeployment --location "$location" --admin-user $SQLAdminLogin --admin-password $SQLAdminLoginPassword
```
Comment (alexdrenea) [inline deployment/Templates/Deploy.ps1:224]: The outputs from these commands can be quite verbose and will pollute the output window making it hard to follow. I usually append `--output none` to all the az cli commands to keep the output clean.

File: deployment/Templates/DeployKV.ps1 [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+# Create RG if not exists
+New-AzResourceGroup -Name $ResourceGroupForDeployment -Location $location -Force
+
+Write-host "📜  Start Deploy resources"
+$WebAppNameService=$WebAppNamePrefix+"AmpSvcPlan"
+$WebAppNameAdmin=$WebAppNamePrefix+"-admin"
+$WebAppNamePortal=$WebAppNamePrefix+"-portal"
+$KeyVault=$WebAppNamePrefix+"-kv"
+$ADApplicationSecretKeyVault="@Microsoft.KeyVault(SecretUri=https://"+$KeyVault+".vault.azure.net/secrets/ADApplicationSecret/)"
+$DefaultConnectionKeyVault="@Microsoft.KeyVault(SecretUri=https://"+$KeyVault+".vault.azure.net/secrets/DefaultConnection/)"
```
Comment (alexdrenea) [inline deployment/Templates/DeployKV.ps1:227]: Due to a bug (?) in either powershell or az cli settings command this style of declaring the key vault reference will not work. (if you look in the app settings config, the value won't be recognized as a keyvault referece as it is missing the closing ')'.

The below method should work instead
```
$ADApplicationSecretKeyVault='"@Microsoft.KeyVault(VaultName={0};SecretName=ADApplicationSecret)"' -f $KeyVault
$DefaultConnectionKeyVault='"@Microsoft.KeyVault(VaultName={0};SecretName=DefaultConnection)"' -f $KeyVault
```

File: deployment/Templates/DeployKV.ps1 [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+$Connection="Data Source=tcp:"+$SQLServerName+".database.windows.net,1433;Initial Catalog=AMPSaaSDB;User Id="+$SQLAdminLogin+"@"+$SQLServerName+".database.windows.net;Password="+$SQLAdminLoginPassword+";"
+
+
+Write-host "Create SQL Server"
+az sql server create --name $SQLServerName --resource-group $ResourceGroupForDeployment --location "$location" --admin-user $SQLAdminLogin --admin-password $SQLAdminLoginPassword
+
+Write-host "Add SQL Server Firewall rules"
+az sql server firewall-rule create --resource-group $ResourceGroupForDeployment --server $SQLServerName -n AllowYourIp --start-ip-address "0.0.0.0" --end-ip-address "0.0.0.0"
+
```
Comment (alexdrenea) [inline deployment/Templates/DeployKV.ps1:236]: When running this locally, the script will fail because the IP of the machine running the script is not part of the firewall.
We could solve this by adding local user's IP to the firewall.
The below works, but happy to try out other things that don't depend on a 3rd party website.

```
$publicIp = (Invoke-WebRequest -uri "http://ifconfig.me/ip").Content
az sql server firewall-rule create --resource-group $ResourceGroupForDeployment --server $SQLServerName -n AllowYourIp --start-ip-address "$publicIp" --end-ip-address "$publicIp"
```

File: deployment/Templates/DeployKV.ps1 [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+$ADApplicationSecretKeyVault="@Microsoft.KeyVault(SecretUri=https://"+$KeyVault+".vault.azure.net/secrets/ADApplicationSecret/)"
+$DefaultConnectionKeyVault="@Microsoft.KeyVault(SecretUri=https://"+$KeyVault+".vault.azure.net/secrets/DefaultConnection/)"
+$Connection="Data Source=tcp:"+$SQLServerName+".database.windows.net,1433;Initial Catalog=AMPSaaSDB;User Id="+$SQLAdminLogin+"@"+$SQLServerName+".database.windows.net;Password="+$SQLAdminLoginPassword+";"
+
+
+Write-host "Create SQL Server"
+az sql server create --name $SQLServerName --resource-group $ResourceGroupForDeployment --location "$location" --admin-user $SQLAdminLogin --admin-password $SQLAdminLoginPassword
+
+Write-host "Add SQL Server Firewall rules"
+az sql server firewall-rule create --resource-group $ResourceGroupForDeployment --server $SQLServerName -n AllowYourIp --start-ip-address "0.0.0.0" --end-ip-address "0.0.0.0"
```
Comment (alexdrenea) [inline deployment/Templates/DeployKV.ps1:235]: [Nit] The rule name should be "AllowAzureIP" not "Your" Ip since this rule is essentially the convention for allowing other Azure services to access this SQL server

File: deployment/Templates/Deploy.ps1 (hunk 1 above)
Comment (msalemcode) [reply deployment/Templates/Deploy.ps1:221]: Added the recommendation

File: deployment/Templates/Deploy.ps1 (hunk 2 above)
Comment (msalemcode) [reply deployment/Templates/Deploy.ps1:268]: Not sure. It make should be SQLAzure type however it was always mysql.

File: deployment/Templates/DeployKV.ps1 (hunk 5 above)
Comment (msalemcode) [reply deployment/Templates/DeployKV.ps1]: we ran this code mutliple time and I re-run it again after reading your comment and still I am not getting any error. after deployment web app is up, It authenticate users and it bring data back from database. it is indication the KV reference is working. Did you try latest code

File: deployment/Templates/DeployKV.ps1 (hunk 6 above)
Comment (msalemcode) [reply deployment/Templates/DeployKV.ps1:243]: In case of running locally. User should modify the rule based on their needs. we have ISV who running it from CI/CD pipeline. I will leave this rule as it is for now and local user should take control over how they want to modify the DB rules.

File: deployment/Templates/DeployKV.ps1 (hunk 7 above)
Comment (msalemcode) [reply deployment/Templates/DeployKV.ps1:242]: updated the rule name

File: deployment/Templates/Deploy.ps1 (hunk 4 above)
Comment (msalemcode) [reply deployment/Templates/Deploy.ps1:218]: when we run into errors during the deployment this verbose does help to track down the error location

**Comment by code4clouds** [comment]
@msalemcode  @alexdrenea  Is this good to go?  Should we close PR #183 ?

File: deployment/Templates/DeployKV.ps1 [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+$ErrorActionPreference = "Stop"
+
+# Checking SQL username
+if($SQLAdminLogin.ToLower() -eq "admin") {
+    Throw "🛑 SQLAdminLogin may not be 'admin'."
+    Exit
+}
+
+# Checking SQL password length
+if($SQLAdminLogin.Length -lt 8) {
```
Comment (alexdrenea) [inline deployment/Templates/DeployKV.ps1:35]: this should be SqlAdminPassword

File: deployment/Templates/DeployKV.ps1 (hunk 5 above)
Comment (alexdrenea) [reply deployment/Templates/DeployKV.ps1]: this could be due to the PowerShell version. 
It certainly does not work in PowerShell 5.1 (and does not really throw an error, it will simply not work after deployment). while this version might work in PowerShell 7, the proposed one will work in either version.

Happy to run a test together and make sure we land on a generic solution that works everywhere.

Comment (code4clouds) [reply deployment/Templates/DeployKV.ps1]: The important version is the one in the Azure Portal cloud shell...  as that is the target version for the install instructions, all other are best effort.

Comment (alexdrenea) [reply deployment/Templates/DeployKV.ps1]: @code4clouds  I understand that, but in my experience, a lot of partners I worked with want to be able to clone and deploy from local. Windows 10 and 11 come preinstalled with PowerShell 5.1 and honestly the script works perfectly with this small exception. I don't see a reason why we could not make the script work on both Ps version :)

**Comment by alexdrenea** [comment]
Reviewed with Magdy today! All looks good! 

Unfortunately it looks like I don't have permissions  / access to formally Approve the review, but +1 from my end!

**Comment by code4clouds** [comment]
@alexdrenea Try again...  I granted triage access. :-)

**Review by alexdrenea — APPROVED** [review]

**Comment by alexdrenea** [comment]
@code4clouds I approved!

I can't link issues to the PR but I think this could also solve:
#88 -> KeyVault is added by default in the deployment, no longer needs extra instructions
#95 -> moving the deployment out of the ARM template resolves the compress-archive issue in PowerShell 5.1
#119 -> temp Resource Group is no longer needed
#229, #235 -> the script uses the new Azure PowerShell commands, and confirmed to work both locally and in CloudShell without entering SubId / TenantId
#236, #293 -> the new script does not create temp storage accounts

**Review by code4clouds — CHANGES_REQUESTED** [review]
Please update the changes requested ASAP to start the final testing.

File: deployment/Templates/DeployKV.ps1 [hunk 9]
```diff
@@ -0,0 +1,331 @@
+﻿# Copyright (c) Microsoft Corporation. All rights reserved.
```
[the remark this answers is not in the document]
Comment (alexdrenea) [reply deployment/Templates/DeployKV.ps1:1]: I support this as well. 
I am of strong opinion that KeyVault should not be optional, and think we can remove the "non-KeyVault" file entirely and just have the one deployment file that comes with KeyVault.

[the remark this answers is not in the document]
Comment (code4clouds) [reply deployment/Templates/DeployKV.ps1:1]: +1 let's remove the non-keyvault file and just have one called Deploy.ps1

**Review by code4clouds — APPROVED** [review]

**Review by code4clouds — APPROVED** [review]

**Review by code4clouds — APPROVED** [review]
